pkg/testutils/testk8sclient.go

```diff
@@ -0,0 +1,80 @@
+package testutils
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	cassdcapi "github.com/k8ssandra/cass-operator/apis/cassandra/v1beta1"
+	k8ssandraapi "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	reaperapi "github.com/k8ssandra/k8ssandra-operator/apis/reaper/v1alpha1"
+	stargateapi "github.com/k8ssandra/k8ssandra-operator/apis/stargate/v1alpha1"
+	promapi "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/stretchr/testify/assert"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type TestK8sClient struct {
+	client.Client
+	TestState      *testing.T
+	UnsafeGetSync  func(ctx context.Context, key client.ObjectKey, obj client.Object) error
+	UnsafeListSync func(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error
+	timeout        time.Duration
+	tick           time.Duration
+}
+
+// Get IS AN ASYNC SAFE client.Get(), it is not the normal `Get()`! Use UnsafeListSync() if you want regular Get()'s behaviour.
+func (my TestK8sClient) Get(ctx context.Context, key client.ObjectKey, obj client.Object) error {
+	var err error
```

Comment:
Beware that the idiom you are using in this Get method and in the List method below is broken: a global variable is being written to inside a goroutine without synchronization. The write is not guaranteed to be seen outside the goroutine. You should pass the error through a channel, or use the `errgroup` package.